.gitlab-ci.yml

```diff
@@ -1,20 +1,111 @@
+###############################################################################
+# Copyright (c) 2016-20, Lawrence Livermore National Security, LLC and Umpire
+# project contributors. See the COPYRIGHT file for details.
 #
-# SPDX-License-Identifier: BSD-3-Clause
-######################################################################################
+# SPDX-License-Identifier: (MIT)
+###############################################################################
+
+###############################################################################
+# General GitLab pipelines configurations for supercomputers and Linux clusters
+# at Lawrence Livermore National Laboratory (LLNL).
+#
+# This entire pipeline is LLNL-specific
+# #############################################################################
+
+# We define the following GitLab pipeline variables:
+#
+# GIT_SUBMODULE_STRATEGY:
+# Tells Gitlab to recursively update the submodules when cloning umpire
+#
+# ALLOC_NAME:
+# On LLNL's ruby, this pipeline creates only one allocation shared among jobs
+# in order to save time and resources. This allocation has to be uniquely named
+# so that we are sure to retrieve it.
+#
+# BUILD_ROOT:
+# The path to the shared resources between all jobs. The BUILD_ROOT is unique to
+# the pipeline, preventing any form of concurrency with other pipelines. This
+# also means that the BUILD_ROOT directory will never be cleaned.
+# TODO: add a clean-up mechanism
 
 variables:
-   GIT_SUBMODULE_STRATEGY: normal
+  GIT_SUBMODULE_STRATEGY: recursive
+  ALLOC_NAME: ${CI_PROJECT_NAME}_ci_${CI_PIPELINE_ID}
+  BUILD_ROOT: ${CI_PROJECT_DIR}
+
+# Normally, stages are blocking in Gitlab. However, using the keyword "needs" we
+# can express dependencies between job that break the ordering of stages, in
+# favor of a DAG.
+# In practice r_*, l_* and b_* stages are independently run and start immediately.
 
 stages:
-   - build
-
-cuda:
-   stage: build
-   tags:
-      - shell
-      - lassen
-   script:
-      - lalloc 1 -W 30 .gitlab/ci/helper.sh
+  - r_allocate_resources
+  - r_build_and_test
+  - r_release_resources
+  - l_build_and_test
+  - c_build_and_test
+#  - multi_project
+
+# This is the rules that drives the activation of "advanced" jobs. All advanced
+# jobs will share this through a template mechanism.
+.advanced_pipeline:
+  rules:
+    - if: '$CI_COMMIT_BRANCH == "main" || $CI_COMMIT_BRANCH == "develop" || $ALL_TARGETS == "ON"' #run only if ...
+
+# These are also templates (.name) that define project specific build commands.
+# If an allocation exist with the name defined in this pipeline, the job will
+# use it (slurm specific).
+.build_toss_3_x86_64_ib_script:
+  script:
+    - echo ${ALLOC_NAME}
+    - export JOBID=$(squeue -h --name=${ALLOC_NAME} --format=%A)
+    - echo ${JOBID}
+    - pwd
+    - ls scripts
+    - ls scripts/gitlab
+    - srun $( [[ -n "${JOBID}" ]] && echo "--jobid=${JOBID}" ) -t 30 -N 1 scripts/gitlab/build_and_test.sh
```

Comment:
What does it mean if JOBID is not defined?

Reply:
JOBID is being defined on line 61...

Reply:
Sorry, I meant what does it mean when it fails the conditional (-n) check?

Reply:
I think the $( ) would evaluate to an empty string at that point, at which point it would get new resources from srun instead of queing the job within the current resource allocation. I think this probably adds some flexibility to whether this is run from within or without an existing resource allocation. (FYI, I copied this from umpire)